### PR TITLE
clientd: spend endpoint and some imporovements

### PR DESCRIPTION
**Pull request by NicolaLS**
### This PR mainly introduces the spend, and fetching functionality for clientd

I also addressed improvement requests and open TODOs from #384 in commits:
- `write json directly to stdout`
- `use macro for success response`
- `add server failure error`

Even though I wanted to keep the PR small I needed the extra stuff to implement the spend endpoint without having to add TODOs/workarounds which I'd need to come back later to.

### **Note for Reviewers**
In case you go through each individual commit, know that I addressed all `//TODO: ..`, in later commits. 
With  exemption of `//TODO: Log event` in the `fetch` function and the `//TODO: Encode coins` in the cli

### CODE REVIEW

**Comment by NicolaLS** [comment]
Just realized it doesn't make a lot of sense for the data to be a `Vec<_>`, because the API doesn't even process `batches` as in a `json-rpc 2.0` spec for example. So there will only ever be one response object per request/response

File: client/clientd/src/main.rs [hunk 1]
```diff
@@ -24,7 +26,8 @@ struct Config {
     workdir: PathBuf,
 }
 struct State {
-    client: Client<UserClientConfig>,
+    client: Arc<Client<UserClientConfig>>,
+    fetch_tx: Sender<()>,
```
Comment (justinmoon) [inline client/clientd/src/main.rs:30]: Let's call this `fetch_sender`.

`fetch_tx` sounds like something that's going to lookup a transaction.

**Comment by justinmoon** [comment]
Why do we need a background thread for fetching?

Seems like it would be better to just fetch in-line and await that future.

**Comment by NicolaLS** [comment]
> Why do we need a background thread for fetching?
> 
> Seems like it would be better to just fetch in-line and await that future.

Later for `reissue` we want to return `200 OK` immediately (not blocking the HTTP request/response) (the reissue can still fail though). Just awaiting the fetch was the first implementation but it got messy like this (spawning threads, and some other stuff I cant remember) so it got changed to the fetch thread (that looked in the DB for active issuances previously but now just waits for signal)

**Comment by NicolaLS** [comment]
> Why do we need a background thread for fetching?
> 
> Seems like it would be better to just fetch in-line and await that future.

Because fetch is something we want do do generally in the background like in `´reissue` and potentially other endpoints. There would be a lot of ```
thread::spawn(async { await fetching })); return OK```
(not correct syntax)
which would be verbose (and also when I was implementing the API as json 2.0 spec there were some problems with that but I cant remamber what exactly)

**Comment by elsirion** [comment]
> Why do we need a background thread for fetching?

Not specifically regarding this PR, but from a high level: confirming a tx takes one epoch, issuing e-cash another. But the issuance is a mere formality as long as the federation isn't compromised. So we can just return success after the tx is confirmed and fetch the e-cash tokens in the background, saving on latency. That's one of the main reasons for `clientd` imo.

**Comment by NicolaLS** [comment]
@justinmoon @elsirion do I need to address something not sure if the fetch thread is ok or not now ?

File: client/clientd/src/bin/clientd-cli.rs [hunk 2]
```diff
@@ -1,10 +1,10 @@
 use anyhow::Result;
 use bitcoin::Transaction;
 use clap::{Parser, Subcommand};
-use clientd::{call, PegInPayload, WaitBlockHeightPayload};
-use fedimint_api::module::__reexports::serde_json;
+use clientd::{call, PegInPayload, SpendPayload, WaitBlockHeightPayload};
+use fedimint_api::{module::__reexports::serde_json, Amount};
```
Comment (elsirion) [inline client/clientd/src/bin/clientd-cli.rs:5]: Just noticed this, could you remove the `__reexports` in a next PR please?

File: client/clientd/src/lib.rs [hunk 3]
```diff
@@ -165,3 +165,25 @@ where
...
+    ($payload:expr) => {
+        {
+       let body = serde_json::json!({
+            "data": $payload
+       });
+
+       Ok(axum::Json(body))
+    }
+    };
+}
```
Comment (elsirion) [inline client/clientd/src/lib.rs:204]: That most likely doesn't *need* to be a macro but ok for now.

**Review by elsirion — APPROVED** [review]
Overall good, let's unblock you here :)

**Pull request merged**